File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/audit/IdentifyUserTest.java
package org.odk.collect.android.feature.formentry.audit;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.apache.commons.csv.CSVRecord;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.junit.runner.RunWith;
import org.odk.collect.android.support.StorageUtils;
import org.odk.collect.android.support.pages.FormEntryPage;
import org.odk.collect.android.support.pages.FormHierarchyPage;
import org.odk.collect.android.support.pages.IdentifyUserPromptPage;
import org.odk.collect.android.support.pages.MainMenuPage;
import org.odk.collect.android.support.rules.CollectTestRule;
import org.odk.collect.android.support.rules.TestRuleChain;

import java.io.IOException;
import java.util.List;

@RunWith(AndroidJUnit4.class)
public class IdentifyUserTest {

    private static final String IDENTIFY_USER_AUDIT_FORM = "identify-user-audit.xml";
    private static final String IDENTIFY_USER_AUDIT_FALSE_FORM = "identify-user-audit-false.xml";

    public CollectTestRule rule = new CollectTestRule();

    @Rule
    public RuleChain copyFormChain = TestRuleChain.chain()
            .around(rule);

    @Test
    public void openingForm_andThenEnteringIdentity_andThenFillingForm_logsUser() throws IOException {
        rule.startAtMainMenu()
                .copyForm(IDENTIFY_USER_AUDIT_FORM)
                .clickFillBlankForm()
                .clickOnFormWithIdentityPrompt("Identify User")
                .enterIdentity("Lucius")
                .clickKeyboardEnter(new FormEntryPage("Identify User"))
                .swipeToEndScreen()
                .clickSaveAndExit();

        List<CSVRecord> auditLog = StorageUtils.getAuditLogForFirstInstance();
        CSVRecord formStartEvent = auditLog.get(0);
        assertThat(formStartEvent.get(0), equalTo("form start"));
        assertThat(formStartEvent.get(4), equalTo("Lucius"));
    }

    @Test
    public void openingSavedForm_andThenEnteringIdentity_andThenFillingForm_logsUser() throws IOException {
        rule.startAtMainMenu()
                .copyForm(IDENTIFY_USER_AUDIT_FORM)
                .clickFillBlankForm()
                .clickOnFormWithIdentityPrompt("Identify User")
                .enterIdentity("Lucius")
                .clickKeyboardEnter(new FormEntryPage("Identify User"))
                .swipeToEndScreen()
                .clickSaveAndExit()

                .clickEditSavedForm()
                .clickOnFormWithIdentityPrompt("Identify User")
                .enterIdentity("Jack")
                .clickKeyboardEnter(new FormHierarchyPage("Identify User"))
                .clickJumpEndButton()
                .clickSaveAndExit();

            List<CSVRecord> auditLog = StorageUtils.getAuditLogForFirstInstance();
            CSVRecord formResumeEvent = auditLog.get(6);
            assertThat(formResumeEvent.get(0), equalTo("form resume"));
            assertThat(formResumeEvent.get(4), equalTo("Jack"));
    }

    @Test
    public void openingForm_andEnteringBlankIdentity_remainsOnIdentityPrompt() {
        rule.startAtMainMenu()
                .copyForm(IDENTIFY_USER_AUDIT_FORM)
                .clickFillBlankForm()
                .clickOnFormWithIdentityPrompt("Identify User")
                .enterIdentity("  ")
                .clickKeyboardEnterWithValidationError();
    }

    @Test
    public void openingForm_andPressingBack_returnsToMainMenu() {
        rule.startAtMainMenu()
                .copyForm(IDENTIFY_USER_AUDIT_FORM)
                .clickFillBlankForm()
                .clickOnFormWithIdentityPrompt("Identify User")
                .closeSoftKeyboard()
                .pressBack(new MainMenuPage());
    }

    @Test
    public void openingForm_andRotating_remainsOnIdentityPrompt() {
        rule.startAtMainMenu()
                .copyForm(IDENTIFY_USER_AUDIT_FORM)
                .clickFillBlankForm()
                .clickOnFormWithIdentityPrompt("Identify User")
                .enterIdentity("Blah")
                .rotateToLandscape(new IdentifyUserPromptPage("Identify User"))
                .assertText("Blah");
    }

    @Test
    public void openingForm_andPressingCloseCross_returnsToMainMenu() {
        rule.startAtMainMenu()
                .copyForm(IDENTIFY_USER_AUDIT_FORM)
                .clickFillBlankForm()
                .clickOnFormWithIdentityPrompt("Identify User")
                .pressClose();
    }

    @Test
    public void openFormWithIdentifyUserFalse_proceedsToForm() {
        rule.startAtMainMenu()
                .copyForm(IDENTIFY_USER_AUDIT_FALSE_FORM)
                .clickFillBlankForm()
                .clickOnForm("Identify User False")
                .swipeToEndScreen()
                .clickSaveAndExit();
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/audit/IdentifyUserTest.kt
package org.odk.collect.android.feature.formentry.audit

import androidx.test.ext.junit.runners.AndroidJUnit4
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.runner.RunWith
import org.odk.collect.android.support.StorageUtils
import org.odk.collect.android.support.StorageUtils.getAuditLogForFirstInstance
import org.odk.collect.android.support.pages.FormEntryPage
import org.odk.collect.android.support.pages.FormHierarchyPage
import org.odk.collect.android.support.pages.IdentifyUserPromptPage
import org.odk.collect.android.support.pages.MainMenuPage
import org.odk.collect.android.support.rules.CollectTestRule
import org.odk.collect.android.support.rules.TestRuleChain
import java.io.File
import java.io.IOException

@RunWith(AndroidJUnit4::class)
class IdentifyUserTest {
    private val rule = CollectTestRule()

    @get:Rule
    val ruleChain: RuleChain = TestRuleChain
        .chain()
        .around(rule)

    @Test
    @Throws(IOException::class)
    fun openingBlankForm_andThenEnteringIdentity_andThenFillingForm_logsUser() {
        rule.startAtMainMenu()
            .copyForm(IDENTIFY_USER_AUDIT_FORM)
            .clickFillBlankForm()
            .clickOnFormWithIdentityPrompt("Identify User")
            .enterIdentity("Lucius")
            .clickKeyboardEnter(FormEntryPage("Identify User"))
            .swipeToEndScreen()
            .clickFinalize()

        val auditLog = getAuditLogForFirstInstance()
        val formStartEvent = auditLog[0]
        assertThat(formStartEvent[0], equalTo("form start"))
        assertThat(formStartEvent[4], equalTo("Lucius"))
    }

    @Test
    @Throws(IOException::class)
    fun openingSavedForm_andThenEnteringIdentity_andThenFillingForm_logsUser() {
        rule.startAtMainMenu()
            .copyForm(IDENTIFY_USER_AUDIT_FORM)
            .clickFillBlankForm()
            .clickOnFormWithIdentityPrompt("Identify User")
            .enterIdentity("Lucius")
            .clickKeyboardEnter(FormEntryPage("Identify User"))
            .swipeToEndScreen()
            .clickSaveAsDraft()
            .clickDrafts()
            .clickOnFormWithIdentityPrompt("Identify User")
            .enterIdentity("Jack")
            .clickKeyboardEnter(FormHierarchyPage("Identify User"))
            .clickJumpEndButton()
            .clickFinalize()

        val auditLog = getAuditLogForFirstInstance()
        val formResumeEvent = auditLog[5]
        assertThat(formResumeEvent[0], equalTo("form resume"))
        assertThat(formResumeEvent[4], equalTo("Jack"))
    }

    @Test
    fun openingBlankForm_andEnteringBlankIdentity_remainsOnIdentityPrompt() {
        rule.startAtMainMenu()
            .copyForm(IDENTIFY_USER_AUDIT_FORM)
            .clickFillBlankForm()
            .clickOnFormWithIdentityPrompt("Identify User")
            .enterIdentity("  ")
            .clickKeyboardEnterWithValidationError()
    }

    @Test
    fun openingBlankForm_andPressingBack_returnsToMainMenuAndDoesNotLeaveAnEmptyInstanceDir() {
        rule.startAtMainMenu()
            .copyForm(IDENTIFY_USER_AUDIT_FORM)
            .clickFillBlankForm()
            .clickOnFormWithIdentityPrompt("Identify User")
            .closeSoftKeyboard()
            .pressBack(MainMenuPage())

        assertThat(File(StorageUtils.getInstancesDirPath()).listFiles().size, equalTo(0))
    }

    @Test
    fun openingBlankForm_andRotating_remainsOnIdentityPrompt() {
        rule.startAtMainMenu()
            .copyForm(IDENTIFY_USER_AUDIT_FORM)
            .clickFillBlankForm()
            .clickOnFormWithIdentityPrompt("Identify User")
            .enterIdentity("Blah")
            .rotateToLandscape(IdentifyUserPromptPage("Identify User"))
            .assertText("Blah")
    }

    @Test
    fun minimizingAndReopeningApp_remainsOnIdentityPrompt() {
        rule.startAtMainMenu()
            .copyForm(IDENTIFY_USER_AUDIT_FORM)
            .clickFillBlankForm()
            .clickOnFormWithIdentityPrompt("Identify User")
            .enterIdentity("Blah")
            .minimizeAndReopenApp(IdentifyUserPromptPage("Identify User"))
            .assertText("Blah")
    }

    @Test
    fun openingBlankForm_andPressingCloseCross_returnsToMainMenuAndDoesNotLeaveAnEmptyInstanceDir() {
        rule.startAtMainMenu()
            .copyForm(IDENTIFY_USER_AUDIT_FORM)
            .clickFillBlankForm()
            .clickOnFormWithIdentityPrompt("Identify User")
            .pressClose()

        assertThat(File(StorageUtils.getInstancesDirPath()).listFiles().size, equalTo(0))
    }

    @Test
    fun openingBlankFormWithIdentifyUserFalse_proceedsToForm() {
        rule.startAtMainMenu()
            .copyForm(IDENTIFY_USER_AUDIT_FALSE_FORM)
            .clickFillBlankForm()
            .clickOnForm("Identify User False")
            .swipeToEndScreen()
            .clickFinalize()
    }

    companion object {
        private const val IDENTIFY_USER_AUDIT_FORM = "identify-user-audit.xml"
        private const val IDENTIFY_USER_AUDIT_FALSE_FORM = "identify-user-audit-false.xml"
    }
}
